metabox/conftest.py

```diff
@@ -0,0 +1,181 @@
+        "--log-lvl", dest="scenario_log_level", choices=Core().levels.keys(),
+        default='WARNING',
+        help="Set the scenario logging level",
+    )
+    parser.addoption(
+        '--do-not-dispose', action='store_true',
+        help="Do not delete LXD containers after the run")
+    parser.addoption(
+        '--debug-machine-setup', action='store_true',
+        help="Turn on verbosity during machine setup. "
+             "Only works with --log TRACE")
+
+
+def pytest_pycollect_makeitem(collector, name, obj):
+    """
+    This function is a pytest hook that allows for the dynamic creation of test
+    items. It creates a pytest item for each Scenario subclass that exists in
+    the Metabox project.
+    """
+    if (
+        inspect.isclass(obj) and issubclass(obj, Scenario) and
+        "Scenario" not in obj.__name__
+    ):
+        variants = []
+        for scn in collector.config.runner.scn_variants:
+            if (
+                "{}.{}".format(
+                    collector.parent.name, collector.path.stem) in scn.name and
+                scn.name.endswith('.' + name)
+            ):
+                variants.append(
+                    MetaboxItem.from_parent(
+                        collector,
+                        name=f"{name} [{scn.mode}]{scn.releases}",
+                        scn=scn)
+                    )
+        return variants
+    elif inspect.isclass(obj) and issubclass(obj, Scenario):
+        return []
+
+
+def pytest_configure(config):
+    """
+    Initialization of the Runner object with the command-line options passed to
+    pytest.
+    """
+    logger = logging.getLogger('ws4py')
+    logger.disabled = True
+    runner = Runner(config.option)
+    runner.collect()
+    config.runner = runner
+
+
+def pytest_collection_finish(session):
+    """
+    This hook is called once all tests have been collected.
+    It sets up the runner with the collected scenarios.
+    """
+    if session.config.option.collectonly:
+        return
+    runner = session.config.runner
+    runner.scn_variants = [i.scn for i in session.items]
+    runner.setup()
+
+
+class MetaboxItem(pytest.Item):
+    """
+    Custom pytest item for running a Scenario.
+    It loads the  appropriate machine configurations based on the scenario
+    variant and mode, sets up the machines, and runs the scenario.
+    It also handles cleanup and reporting of any failures that occur during the
+    run.
+    """
+    def __init__(self, *, scn, **kwargs):
+        super().__init__(**kwargs)
+        self.scn = scn
+
+    def __repr__(self) -> str:
+        return "<Scenario {}>".format(getattr(self, "name", None))
+
+    def _load(self, mode, release_alias):
+        config = self.config.runner.config[mode].copy()
+        config['alias'] = release_alias
+        config['role'] = mode
+        return self.config.runner.machine_provider.get_machine_by_config(
+            MachineConfig(mode, config))
+
+    def runtest(self):
```

Comment:
this `runtest()` method and the `_load()` method above look a lot like the `run()` and `_load()` methods from [`metabox.core.runner`](https://github.com/canonical/checkbox/blob/metabox-pytest/metabox/metabox/core/runner.py#L180-L240). There might be a chance for refactoring here...